core/schemas/__init__.py

```diff
@@ -19,30 +20,76 @@
+    pattern_matcher = re.compile(type_pattern)
+    for schema_file in Path(__file__).parent.glob(f"{schema_root_type}/**/*.py"):
+        if schema_file.stem == "__init__":
             continue
-        logger.info(f"Registering entity type {entity_file.stem}")
-        if entity_file.parent.stem == "entities":
-            module_name = f"core.schemas.entities.{entity_file.stem}"
-        elif entity_file.parent.stem == "private":
-            module_name = f"core.schemas.entities.private.{entity_file.stem}"
-        enum_value = entity_file.stem.replace("_", "-")
-        if entity_file.stem not in entity.EntityType.__members__:
-            aenum.extend_enum(entity.EntityType, entity_file.stem, enum_value)
-        modules[module_name] = enum_value
-    entity.TYPE_MAPPING = {"entity": entity.Entity, "entities": entity.Entity}
-    for module_name, enum_value in modules.items():
+        if schema_file.parent.stem == schema_root_type:
+            module_name = f"core.schemas.{schema_root_type}.{schema_file.stem}"
+        elif schema_file.parent.stem == "private":
+            module_name = f"core.schemas.{schema_root_type}.private.{schema_file.stem}"
+        with open(schema_file, "r") as f:
+            content = f.read()
+        for schema_type in pattern_matcher.findall(content):
+            if schema_type not in schema_enum.__members__:
+                logger.debug(
+                    f"Adding observable type <{schema_type}> to {schema_enum.__name__} enum"
+                )
+                if schema_root_type == "entities":
+                    aenum.extend_enum(
+                        schema_enum, schema_type, schema_type.replace("_", "-")
+                    )
+                else:
+                    aenum.extend_enum(schema_enum, schema_type, schema_type)
+                modules.add(module_name)
+            else:
+                logger.warning(
+                    f"Observable type {schema_type} defined in <{module_name}> already exists"
+                )
+    return modules
+
+
+def register_schema_classes(base_module, base_class, modules: set, type_mapping: dict):
+    """
+    Register the schemas from the implementation files
+    :param base_module: The schema root type to work with
+    :param base_class: base class that the schema class should inherit from
+    :param modules: The modules to register
+    :param type_mapping: schema type mapping to update
+    """
+    logger.info(f"Registering {base_module.__name__} classes")
+    for module_name in modules:
         module = importlib.import_module(module_name)
         for _, obj in inspect.getmembers(module, inspect.isclass):
-            if issubclass(obj, entity.Entity):
-                entity.TYPE_MAPPING[enum_value] = obj
-                setattr(entity, obj.__name__, obj)
+            if issubclass(obj, base_class):
+                if "type" in obj.model_fields:
+                    obs_type = obj.model_fields["type"].default.value
```

Comment:
Can't we extend the type enums here instead of doing regex matching on the python file? The latter will break if we ever change our type hinting notation.